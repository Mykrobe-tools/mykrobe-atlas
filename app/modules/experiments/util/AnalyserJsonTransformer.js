/* @flow */

import { isArray } from 'makeandship-js-common/src/utils/is';

import susceptibilityTransformer from './transformers/susceptibility';
import speciesTransformer from './transformers/species';

export type AnalyserJsonTransformerResult = {
  lineage: Array<string>,
  species: Array<string>,
  speciesAndLineageString: string,
  hasSpecies: boolean,
  hasResistance: boolean,
  susceptible: Array<string>,
  resistant: Array<string>,
  inconclusive: Array<string>,
  positive: Array<string>,
  negative: Array<string>,
  inducible: Array<string>,
  evidence: any,
  drugsResistance: {
    mdr: boolean,
    xdr: boolean,
  },
  samples?: any,
  tree?: any,
  error?: any,
  distance?: any,
};

class AnalyserJsonTransformer {
  transform(jsonString: string) {
    return this.stringToJson(jsonString);
  }

  stringToJson(string: string): Promise<{ json: Object, transformed: Object }> {
    return new Promise((resolve, reject) => {
      let json;

      try {
        json = JSON.parse(string);
      } catch (error) {
        return reject(`Error reading json: ${error.message}`);
      }

      if (!json) {
        return reject(`Error reading json`);
      }

      try {
        const transformed = this.transformModel(json);
        return resolve({ json, transformed });
      } catch (error) {
        return reject(error);
      }
    });
  }

  transformModel(sourceModel: Object): Object {
    if (sourceModel.results) {
      // web
      let sampleModel;
      if (isArray(sourceModel.results)) {
        // if it's an array, just do the first one for now
        sampleModel = sourceModel.results[0];
      } else {
        // use the explicitly defined predictor result
        sampleModel = sourceModel.results['predictor'];
      }
      // console.log('sampleModel', JSON.stringify(sampleModel, null, 2));
      const transformedSampleModel = this.transformSampleModel(sampleModel);
      return transformedSampleModel;
    }
    if (sourceModel.snpDistance) {
      // just do the first one for now
      const sampleIds = Object.keys(sourceModel.snpDistance.newick);
      const sampleId = sampleIds[0];

      const sampleModel = sourceModel.snpDistance.newick[sampleId];

      const transformedSampleModel = this.transformSampleModel(
        sampleModel,
        sourceModel.geoDistance.experiments
      );

      return transformedSampleModel;
    } else {
      // only one sample from Predictor
      const sampleIds = Object.keys(sourceModel);
      if (sampleIds.length === 1) {
        const sampleId = sampleIds[0];
        const sampleModel = sourceModel[sampleId];
        const transformedSampleModel = this.transformSampleModel(sampleModel);
        return transformedSampleModel;
      }
      // Already unwrapped
      const transformedSampleModel = this.transformSampleModel(sourceModel);
      return transformedSampleModel;
    }
  }

  transformSampleModel(
    sourceModel: Object,
    relatedModels: ?Array<Object>
  ): AnalyserJsonTransformerResult {
    let transformed: AnalyserJsonTransformerResult = {};
    // return empty object if there is no data to parse
    if (
      !sourceModel ||
      typeof sourceModel !== 'object' ||
      Object.keys(sourceModel).length === 0
    ) {
      return transformed;
    }
    // check basic requirements
    if (!sourceModel.phylogenetics) {
      if (IS_ELECTRON) {
        // in Electron we allow user opening json file, so show an error
        throw new Error('Unsupported sample json format');
      }
      return transformed;
    }

    // can we display anything?

    transformed = {
      ...transformed,
      ...this.transformFlags(sourceModel),
    };

    const { hasSpecies, hasResistance } = transformed;

    if (!hasSpecies && !hasResistance) {
      const error = `This sample does not appear to contain any Mycobacterial data (or it is amplicon data, which is not supported), and therefore the predictor does not give susceptibility predictions`;
      if (IS_ELECTRON) {
        throw new Error(error);
      }
      return { error };
    }

    // style species for display

    transformed = {
      ...transformed,
      ...speciesTransformer(sourceModel),
    };

    // return now if there is no resistance to show

    if (!hasResistance) {
      return transformed;
    }

    let key;
    let value;

    const susceptibilityModel = sourceModel['susceptibility'];

    const genotypeModel =
      sourceModel['genotype_model'] || sourceModel['genotypeModel'];

    const susceptibility = susceptibilityTransformer(
      susceptibilityModel,
      genotypeModel
    );
    transformed = {
      ...transformed,
      ...susceptibility,
    };

    transformed.positive = [];
    transformed.negative = [];

    const virulenceModel =
      sourceModel['virulence_toxins'] || sourceModel['virulenceToxins'];

    if (virulenceModel) {
      for (key in virulenceModel) {
        value = virulenceModel[key].toUpperCase();
        if (value === 'POSITIVE') {
          transformed.positive.push(key);
        } else if (value === 'NEGATIVE') {
          transformed.negative.push(key);
        }
      }
    }

    let drugsResistance = {
      mdr: false,
      xdr: false,
    };

    /*
    https://www.who.int/tb/areas-of-work/drug-resistant-tb/xdr-tb-faq/en/
    XDR-TB involves resistance to the two most powerful anti-TB drugs, isoniazid and rifampicin, also known as multidrug-resistance (MDR-TB), in addition to resistance to any of the fluoroquinolones (such as levofloxacin or moxifloxacin) and to at least one of the three injectable second-line drugs (amikacin, capreomycin or kanamycin).
    */

    // Check for multidrug-resistance (MDR)

    const mdr =
      transformed.resistant.includes('Isoniazid') &&
      transformed.resistant.includes('Rifampicin');

    if (mdr) {
      drugsResistance.mdr = true;

      // Check for extensively drug-resistant (XDR)

      const fluoroquinolonesResistant =
        transformed.resistant.includes('Ofloxacin') ||
        transformed.resistant.includes('Moxifloxacin') ||
        transformed.resistant.includes('Ciprofloxacin');

      if (fluoroquinolonesResistant) {
        if (
          transformed.resistant.includes('Amikacin') ||
          transformed.resistant.includes('Kanamycin') ||
          transformed.resistant.includes('Capreomycin')
        ) {
          drugsResistance.xdr = true;
        }
      }
    }

    transformed.drugsResistance = drugsResistance;

    // tree and neighbours
    if (sourceModel.neighbours && relatedModels) {
      let neighbourKeys = Object.keys(sourceModel.neighbours);
      let samples = {};
      for (let i = 0; i < 2; i++) {
        const neighbour = sourceModel.neighbours[neighbourKeys[i]];
        let keys = Object.keys(neighbour);
        let neighbourSampleModel = relatedModels[i];
        let sampleId: string = keys[0];
        neighbourSampleModel.id = sampleId;
        samples[sampleId] = neighbourSampleModel;
      }
      transformed.samples = samples;

      transformed.tree = sourceModel.tree;
    }

    return transformed;
  }

  /*
    {
      "phylogenetics": {
      "lineage": {
        "Beijing_East_Asia": {
          "percent_coverage": 100,
          "median_depth": 36
        }
      },
      "sub_complex": {
        "Unknown": {
          "percent_coverage": -1,
          "median_depth": -1
        }
      },
      "phylo_group": {
        "Mycobacterium_tuberculosis_complex": {
          "percent_coverage": 99.496,
          "median_depth": 38
        }
      },
      "species": {
        "Mycobacterium_tuberculosis": {
          "percent_coverage": 98.454,
          "median_depth": 34
        }
      }
    },
  */

  /*
  test if the phylo_group is either:
  1) Non_tuberculosis_mycobacterium_complex, in which case only the species will be shown an no AMR predictions
  2) Mycobacterium_tuberculosis_complex, in which case both species and AMR prediction will be performed.
  */

  transformFlags(
    sourceModel: Object
  ): { hasSpecies: boolean, hasResistance: boolean } {
    let hasSpecies = false;
    let hasResistance = false;

    if (sourceModel.phylogenetics && sourceModel.phylogenetics.phylo_group) {
      const phyloGroup = Object.keys(sourceModel.phylogenetics.phylo_group);
      if (phyloGroup.includes('Non_tuberculosis_mycobacterium_complex')) {
        hasSpecies = true;
      }
      if (phyloGroup.includes('Mycobacterium_tuberculosis_complex')) {
        hasSpecies = true;
        hasResistance = true;
      }
    }

    return { hasSpecies, hasResistance };
  }
}

export default AnalyserJsonTransformer;
