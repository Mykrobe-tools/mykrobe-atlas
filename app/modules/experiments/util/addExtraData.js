/* @flow */

export default (json: Object) => {
  const testData = require('../__fixtures__/experiment.json').data;
  if (!json.location) {
    json.location = testData.location;
  }
  if (!json.geoDistance) {
    json.geoDistance = testData.geoDistance;
  }
  if (json.geoDistance.experiments.length === 0) {
    json.geoDistance.experiments = testData.geoDistance.experiments;
  }
  if (!json.snpDistance) {
    json.snpDistance = testData.snpDistance;
  }
  if (!json.snpDistance.newick) {
    json.snpDistance.newick = testData.snpDistance.newick;
  }
  return json;
};
