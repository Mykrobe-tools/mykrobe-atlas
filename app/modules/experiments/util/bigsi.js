/* @flow */

import _get from 'lodash.get';

export const mapTypeToSearch = {
  sequence: 'Sequence search',
  'protein-variant': 'Protein variant search',
  'dna-variant': 'DNA variant search',
};

export const descriptionForBigsi = (bigsi) => {
  const query = _get(bigsi, 'query.seq');
  const type = _get(bigsi, 'type');
  const search = mapTypeToSearch[type] || 'Search';
  return query ? `${search} for ${query}` : search;
};
