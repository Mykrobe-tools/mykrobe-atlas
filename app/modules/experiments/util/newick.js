/* @flow */

import memoizeOne from 'memoize-one';

export const getNewickTokens = memoizeOne(newick =>
  newick.split(/\s*(;|\(|\)|,|:)\s*/)
);

export const newickContainsNodeId = (newick, nodeId) => {
  const tokens = getNewickTokens(newick);
  return tokens && tokens.includes(nodeId);
};
