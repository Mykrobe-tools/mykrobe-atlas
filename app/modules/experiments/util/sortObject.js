/* @flow */

const sortObject = (o: Object) => {
  let sorted = {};
  let key;
  let a = [];

  for (key in o) {
    if (o.hasOwnProperty(key)) {
      a.push(key);
    }
  }

  a.sort();

  for (key = 0; key < a.length; key++) {
    sorted[a[key]] = o[a[key]];
  }
  return sorted;
};

export default sortObject;
