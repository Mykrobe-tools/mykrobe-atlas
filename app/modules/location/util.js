/* @flow */

// rather than use query-string, these functions preserve order of keys
// this is necessary when dynamically adding and removing filters
// so that the ui will maintain order

export const parseQuery = (query: string): any => {
  const searchParams = new URLSearchParams(query);
  const parsed = {};
  for (let param of searchParams) {
    const key = param[0];
    const value = param[1];
    parsed[key] = value;
  }
  return parsed;
};

export const stringifyQuery = (query: any): string => {
  var searchParams = new URLSearchParams('');
  const queryKeys = Object.keys(query);
  for (let i = 0; i < queryKeys.length; i++) {
    const key = queryKeys[i];
    const value = query[key];
    searchParams.set(key, value);
  }
  return searchParams.toString();
};
