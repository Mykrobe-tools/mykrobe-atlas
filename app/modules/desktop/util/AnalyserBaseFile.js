/* @flow */

import EventEmitter from 'events';

// TODO: we can remove the transformer - it is created on demand by a redux selector
import AnalyserJsonTransformer from '../../experiments/util/AnalyserJsonTransformer';
import * as APIConstants from '../../../constants/APIConstants';

class AnalyserBaseFile extends EventEmitter {
  extensionForFileName(fileName: string) {
    const extension = fileName.substr(fileName.lastIndexOf('.'));
    return extension.toLowerCase();
  }

  analyseFile(filePaths: Array<string>, id: string = ''): AnalyserBaseFile {
    this.cancel();
    for (let i = 0; i < filePaths.length; i++) {
      const filePath = filePaths[i];
      const extension = this.extensionForFileName(filePath);
      if (extension === '.json') {
        return this.analyseJsonFile(filePath);
      }
      if (
        !APIConstants.API_SAMPLE_EXTENSIONS_ARRAY_WITH_DOTS.includes(extension)
      ) {
        const acceptable = APIConstants.API_SAMPLE_EXTENSIONS_ARRAY_WITH_DOTS.join(
          ', '
        );
        this.failWithError(
          `Can only process files with extension: ${acceptable} - not ${extension}`
        );
        return this;
      }
      if (filePath.indexOf(' ') > 0) {
        this.failWithError(
          'Mykrobe does not currently work with files or paths containing spaces'
        );
        return this;
      }
    }
    return this.analyseBinaryFile(filePaths, id);
  }

  failWithError(err: string | Error) {
    let message = err;
    if (err.message) {
      message = err.message;
    }
    setTimeout(() => {
      this.emit('error', {
        description: message,
      });
    }, 0);
  }

  doneWithJsonString(jsonString: string) {
    // const fs = require('fs');
    // fs.writeFileSync('doneWithJsonString.json', jsonString);
    const transformer = new AnalyserJsonTransformer();
    transformer
      .transform(jsonString)
      .then(result => {
        // const { json, transformed } = result;
        // console.log('json', json);
        // console.log('transformed', transformed);
        this.emit('done', result);
      })
      .catch(err => {
        this.failWithError(err);
      });
  }

  analyseJsonFile(file: File | string): AnalyserBaseFile {
    console.log('analyseJsonFile', file);
    return this;
  }

  analyseBinaryFile(filePaths: Array<string>): AnalyserBaseFile {
    console.log('analyseBinaryFile', filePaths);
    return this;
  }

  analyseRemoteFile(file: File | string): AnalyserBaseFile {
    console.log('analyseRemoteFile', file);
    return this;
  }

  cancel(): void {}
}

export default AnalyserBaseFile;
