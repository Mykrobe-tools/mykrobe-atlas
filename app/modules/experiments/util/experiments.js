/* @flow */

import _get from 'lodash.get';

import { newickContainsNodeId } from './newick';

export const experimentsInTree = (
  experimentsTree,
  experiments,
  inTree = true
) => {
  console.log('experimentsInTree', { experimentsTree, experiments });
  return (
    experiments &&
    experiments.filter((experiment) => {
      const leafId = experiment?.leafId;
      let contains = false;
      if (leafId) {
        contains = newickContainsNodeId(experimentsTree, leafId);
      }
      if (contains) {
        return inTree;
      }
      return !inTree;
    })
  );
};

export const experimentsWithGeolocation = (
  experiments,
  withGeolocation = true
) => {
  return (
    experiments &&
    experiments.filter((experiment) => {
      const longitudeIsolate = _get(
        experiment,
        'metadata.sample.longitudeIsolate'
      );
      const latitudeIsolate = _get(
        experiment,
        'metadata.sample.latitudeIsolate'
      );
      if (longitudeIsolate && latitudeIsolate) {
        return withGeolocation;
      }
      return !withGeolocation;
    })
  );
};
