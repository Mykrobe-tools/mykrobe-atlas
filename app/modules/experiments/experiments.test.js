/* @flow */

import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import nock from 'nock';

import { ensureEnv, env } from 'makeandship-js-common/src/util';
const API_URL = ensureEnv(env.API_URL);

import reducer, { requestExperiments } from './experiments';

const createMockStore = configureMockStore([thunk]);
const data = require('../../../test/__fixtures__/api/experiments.json');

describe('experiments module', () => {
  const store = createMockStore([]);
  let mockState;

  beforeEach(() => {
    store.clearActions();
    nock.cleanAll();
  });

  it('should handle "requestExperiments" action', async () => {
    nock(API_URL)
      .get('/experiments/search')
      .query(true)
      .reply(200, data);
    store.dispatch(requestExperiments());
    const dispatchedActions = store.getActions();
    dispatchedActions.forEach(dispatchedAction => {
      mockState = reducer(mockState, dispatchedAction);
    });
    expect(dispatchedActions).toMatchSnapshot();
    expect(mockState).toMatchSnapshot();
  });
});
