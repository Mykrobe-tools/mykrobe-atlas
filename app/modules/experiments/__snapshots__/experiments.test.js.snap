// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`experiments module should handle "requestExperiments" action 1`] = `
Array [
  Object {
    "type": "experiments/experiments/REQUEST_EXPERIMENTS",
  },
  Object {
    "payload": Object {
      "results": Array [
        Object {
          "collected": "2017-03-01T21:42:46.817Z",
          "file": "payment_grocery_invoice.wax",
          "geoDistance": Object {
            "analysed": "2016-10-15T22:19:58.241Z",
            "engine": "numquam",
            "experiments": Array [],
            "id": "591ab838d1f7fd000e50851b",
            "version": "v40896",
          },
          "id": "591ab838d1f7fd000e50851a",
          "jaccardIndex": Object {
            "analysed": "2016-07-16T04:52:03.311Z",
            "engine": "et",
            "experiments": Array [],
            "id": "591ab838d1f7fd000e50851d",
            "version": "v86673",
          },
          "location": Object {
            "id": "591ab838d1f7fd000e50851e",
            "lat": -7.9776,
            "lng": -105.9862,
            "name": "West Ethaport",
          },
          "metadata": Object {
            "anatomicalOrigin": "Other known site",
            "art": "false",
            "bmi": 21.92948782577033,
            "cityIsolate": "West Walterborough",
            "collectionDate": "2016-10-24T10:04:46.023Z",
            "continuationStartDate": "2016-06-15T01:24:40.492Z",
            "continuationStopDate": "2016-07-31T11:54:48.442Z",
            "countryIsolate": "Saint Lucia",
            "countryOfBirth": "Tajikistan",
            "currentTbinformation": true,
            "dateArrived": "2016-08-01T02:22:44.142Z",
            "dateOfDeath": "2016-06-23T08:46:28.140Z",
            "diabetic": "true",
            "genderAtBirth": "female",
            "genexpert": "RIF resistant",
            "hain": "INH/RIF test",
            "hainAm": "AM inconclusive",
            "hainEth": "ETH test failed",
            "hainFl": "FL test failed",
            "hainInh": "INH resistant",
            "hainRif": "RIF resistant",
            "hivStatus": "true",
            "homeless": "true",
            "id": "591ab837d1f7fd000e508437",
            "imprisoned": "false",
            "injectingDrugUse": "false",
            "intensiveStartDate": "2016-06-05T10:43:58.807Z",
            "intensiveStopDate": "2017-01-04T00:51:38.707Z",
            "isolateId": "d330573a-64e5-4511-bc2a-3ab1dbdd898c",
            "labId": "76a923a7-32e8-45ea-bd03-f3d448fa26a0",
            "nonStandardTreatment": "No",
            "otherGenotypeInformation": false,
            "patientAge": 25,
            "patientId": "8af73cc8-7314-4e65-bd72-93b0ae93b471",
            "phenotypeInformationFirstLineDrugs": false,
            "phenotypeInformationOtherDrugs": true,
            "previousTbinformation": true,
            "priorTreatmentDate": "2016-08-07T01:36:49.818Z",
            "prospectiveIsolate": true,
            "recentMdrTb": "Yes",
            "siteId": "98fa2136-28e8-44ce-bf7d-3382f2a7afe1",
            "smear": "++",
            "smoker": "false",
            "sputumCultureConversion": "Sputum culture negative at 2-3 months",
            "sputumSmearConversion": "Not known or not done",
            "startProgrammaticContinuationTreatment": "Yes",
            "startProgrammaticTreatment": false,
            "tbProphylaxis": "Not known",
            "tbProphylaxisDate": "2016-10-18T09:41:31.618Z",
            "wgsPlatform": "NextSeq",
            "wgsPlatformOther": "ullam",
            "whoOutcomeCategory": "Not known",
          },
          "organisation": Object {
            "id": "591ab833d1f7fd000e5083d3",
            "name": "Casper Group",
          },
          "owner": Object {
            "email": "Kendrick.Stamm@yahoo.com",
            "firstname": "Willy",
            "id": "591ab833d1f7fd000e508405",
            "lastname": "Hackett",
            "phone": "957-443-3870 x5326",
          },
          "snpDistance": Object {
            "analysed": "2017-02-25T13:44:01.747Z",
            "engine": "quia",
            "experiments": Array [],
            "id": "591ab838d1f7fd000e50851c",
            "version": "v93582",
          },
          "uploaded": "2016-11-24T23:48:42.488Z",
        },
        Object {
          "collected": "2017-05-16T03:17:59.282Z",
          "file": "configuration.com",
          "geoDistance": Object {
            "analysed": "2017-03-10T18:09:49.306Z",
            "engine": "nihil",
            "experiments": Array [],
            "id": "591ab838d1f7fd000e5084f8",
            "version": "v41776",
          },
          "id": "591ab838d1f7fd000e5084f7",
          "jaccardIndex": Object {
            "analysed": "2016-10-29T06:54:31.060Z",
            "engine": "eligendi",
            "experiments": Array [],
            "id": "591ab838d1f7fd000e5084fa",
            "version": "v82933",
          },
          "location": Object {
            "id": "591ab838d1f7fd000e5084fb",
            "lat": 15.892,
            "lng": 61.2435,
            "name": "Mohamedberg",
          },
          "metadata": Object {
            "anatomicalOrigin": "Other known site",
            "art": "true",
            "bmi": 23.7909730169333,
            "cityIsolate": "Howeview",
            "collectionDate": "2016-12-13T15:56:17.124Z",
            "continuationStartDate": "2017-05-07T18:53:06.144Z",
            "continuationStopDate": "2016-09-23T06:31:23.538Z",
            "countryIsolate": "South Africa",
            "countryOfBirth": "Saudi Arabia",
            "currentTbinformation": true,
            "dateArrived": "2016-12-23T21:57:37.274Z",
            "dateOfDeath": "2016-12-06T00:01:06.750Z",
            "diabetic": "true",
            "genderAtBirth": "male",
            "genexpert": "Not tested",
            "hain": "Fluoroquinolone / aminoglycoside / ethambutol test",
            "hainAm": "AM test failed",
            "hainEth": "ETH resistant",
            "hainFl": "FL sensitive ",
            "hainInh": "INH resistant",
            "hainRif": "RIF inconclusive",
            "hivStatus": "true",
            "homeless": "true",
            "id": "591ab837d1f7fd000e508427",
            "imprisoned": "true",
            "injectingDrugUse": "false",
            "intensiveStartDate": "2017-05-07T09:07:17.599Z",
            "intensiveStopDate": "2017-05-11T22:31:01.111Z",
            "isolateId": "30ccea50-3c19-4c7a-b0a7-07bc791520f2",
            "labId": "9eb49874-5cf6-4c79-a55d-2a2155bf9459",
            "nonStandardTreatment": "Yes",
            "otherGenotypeInformation": false,
            "patientAge": 60,
            "patientId": "df405f1b-30bf-4b36-bdb8-763f6de8bed2",
            "phenotypeInformationFirstLineDrugs": false,
            "phenotypeInformationOtherDrugs": false,
            "previousTbinformation": false,
            "priorTreatmentDate": "2017-05-12T20:19:53.138Z",
            "prospectiveIsolate": false,
            "recentMdrTb": "Not known",
            "siteId": "049fb71d-8157-412e-a289-e5c74cf3d1ef",
            "smear": "+",
            "smoker": "false",
            "sputumCultureConversion": "Sputum culture positive at 2-3 months",
            "sputumSmearConversion": "Not known or not done",
            "startProgrammaticContinuationTreatment": "Not known",
            "startProgrammaticTreatment": true,
            "tbProphylaxis": "Not known",
            "tbProphylaxisDate": "2016-10-17T22:25:29.319Z",
            "wgsPlatform": "Other",
            "wgsPlatformOther": "in",
            "whoOutcomeCategory": "Lost to follow-up or defaulted",
          },
          "organisation": Object {
            "id": "591ab833d1f7fd000e5083c3",
            "name": "Ward - Auer",
          },
          "owner": Object {
            "email": "Anya46@yahoo.com",
            "firstname": "Saige",
            "id": "591ab833d1f7fd000e5083f5",
            "lastname": "Marks",
            "phone": "(773) 716-1334",
          },
          "snpDistance": Object {
            "analysed": "2017-01-21T01:52:11.092Z",
            "engine": "enim",
            "experiments": Array [],
            "id": "591ab838d1f7fd000e5084f9",
            "version": "v80049",
          },
          "uploaded": "2016-10-26T17:12:17.333Z",
        },
      ],
      "summary": Object {
        "hits": 2,
      },
    },
    "type": "experiments/experiments/REQUEST_EXPERIMENTS_SUCCESS",
  },
]
`;

exports[`experiments module should handle "requestExperiments" action 2`] = `
Object {
  "isFetching": false,
  "samples": Array [
    Object {
      "collected": "2017-03-01T21:42:46.817Z",
      "file": "payment_grocery_invoice.wax",
      "geoDistance": Object {
        "analysed": "2016-10-15T22:19:58.241Z",
        "engine": "numquam",
        "experiments": Array [],
        "id": "591ab838d1f7fd000e50851b",
        "version": "v40896",
      },
      "id": "591ab838d1f7fd000e50851a",
      "jaccardIndex": Object {
        "analysed": "2016-07-16T04:52:03.311Z",
        "engine": "et",
        "experiments": Array [],
        "id": "591ab838d1f7fd000e50851d",
        "version": "v86673",
      },
      "location": Object {
        "id": "591ab838d1f7fd000e50851e",
        "lat": -7.9776,
        "lng": -105.9862,
        "name": "West Ethaport",
      },
      "metadata": Object {
        "anatomicalOrigin": "Other known site",
        "art": "false",
        "bmi": 21.92948782577033,
        "cityIsolate": "West Walterborough",
        "collectionDate": "2016-10-24T10:04:46.023Z",
        "continuationStartDate": "2016-06-15T01:24:40.492Z",
        "continuationStopDate": "2016-07-31T11:54:48.442Z",
        "countryIsolate": "Saint Lucia",
        "countryOfBirth": "Tajikistan",
        "currentTbinformation": true,
        "dateArrived": "2016-08-01T02:22:44.142Z",
        "dateOfDeath": "2016-06-23T08:46:28.140Z",
        "diabetic": "true",
        "genderAtBirth": "female",
        "genexpert": "RIF resistant",
        "hain": "INH/RIF test",
        "hainAm": "AM inconclusive",
        "hainEth": "ETH test failed",
        "hainFl": "FL test failed",
        "hainInh": "INH resistant",
        "hainRif": "RIF resistant",
        "hivStatus": "true",
        "homeless": "true",
        "id": "591ab837d1f7fd000e508437",
        "imprisoned": "false",
        "injectingDrugUse": "false",
        "intensiveStartDate": "2016-06-05T10:43:58.807Z",
        "intensiveStopDate": "2017-01-04T00:51:38.707Z",
        "isolateId": "d330573a-64e5-4511-bc2a-3ab1dbdd898c",
        "labId": "76a923a7-32e8-45ea-bd03-f3d448fa26a0",
        "nonStandardTreatment": "No",
        "otherGenotypeInformation": false,
        "patientAge": 25,
        "patientId": "8af73cc8-7314-4e65-bd72-93b0ae93b471",
        "phenotypeInformationFirstLineDrugs": false,
        "phenotypeInformationOtherDrugs": true,
        "previousTbinformation": true,
        "priorTreatmentDate": "2016-08-07T01:36:49.818Z",
        "prospectiveIsolate": true,
        "recentMdrTb": "Yes",
        "siteId": "98fa2136-28e8-44ce-bf7d-3382f2a7afe1",
        "smear": "++",
        "smoker": "false",
        "sputumCultureConversion": "Sputum culture negative at 2-3 months",
        "sputumSmearConversion": "Not known or not done",
        "startProgrammaticContinuationTreatment": "Yes",
        "startProgrammaticTreatment": false,
        "tbProphylaxis": "Not known",
        "tbProphylaxisDate": "2016-10-18T09:41:31.618Z",
        "wgsPlatform": "NextSeq",
        "wgsPlatformOther": "ullam",
        "whoOutcomeCategory": "Not known",
      },
      "organisation": Object {
        "id": "591ab833d1f7fd000e5083d3",
        "name": "Casper Group",
      },
      "owner": Object {
        "email": "Kendrick.Stamm@yahoo.com",
        "firstname": "Willy",
        "id": "591ab833d1f7fd000e508405",
        "lastname": "Hackett",
        "phone": "957-443-3870 x5326",
      },
      "snpDistance": Object {
        "analysed": "2017-02-25T13:44:01.747Z",
        "engine": "quia",
        "experiments": Array [],
        "id": "591ab838d1f7fd000e50851c",
        "version": "v93582",
      },
      "uploaded": "2016-11-24T23:48:42.488Z",
    },
    Object {
      "collected": "2017-05-16T03:17:59.282Z",
      "file": "configuration.com",
      "geoDistance": Object {
        "analysed": "2017-03-10T18:09:49.306Z",
        "engine": "nihil",
        "experiments": Array [],
        "id": "591ab838d1f7fd000e5084f8",
        "version": "v41776",
      },
      "id": "591ab838d1f7fd000e5084f7",
      "jaccardIndex": Object {
        "analysed": "2016-10-29T06:54:31.060Z",
        "engine": "eligendi",
        "experiments": Array [],
        "id": "591ab838d1f7fd000e5084fa",
        "version": "v82933",
      },
      "location": Object {
        "id": "591ab838d1f7fd000e5084fb",
        "lat": 15.892,
        "lng": 61.2435,
        "name": "Mohamedberg",
      },
      "metadata": Object {
        "anatomicalOrigin": "Other known site",
        "art": "true",
        "bmi": 23.7909730169333,
        "cityIsolate": "Howeview",
        "collectionDate": "2016-12-13T15:56:17.124Z",
        "continuationStartDate": "2017-05-07T18:53:06.144Z",
        "continuationStopDate": "2016-09-23T06:31:23.538Z",
        "countryIsolate": "South Africa",
        "countryOfBirth": "Saudi Arabia",
        "currentTbinformation": true,
        "dateArrived": "2016-12-23T21:57:37.274Z",
        "dateOfDeath": "2016-12-06T00:01:06.750Z",
        "diabetic": "true",
        "genderAtBirth": "male",
        "genexpert": "Not tested",
        "hain": "Fluoroquinolone / aminoglycoside / ethambutol test",
        "hainAm": "AM test failed",
        "hainEth": "ETH resistant",
        "hainFl": "FL sensitive ",
        "hainInh": "INH resistant",
        "hainRif": "RIF inconclusive",
        "hivStatus": "true",
        "homeless": "true",
        "id": "591ab837d1f7fd000e508427",
        "imprisoned": "true",
        "injectingDrugUse": "false",
        "intensiveStartDate": "2017-05-07T09:07:17.599Z",
        "intensiveStopDate": "2017-05-11T22:31:01.111Z",
        "isolateId": "30ccea50-3c19-4c7a-b0a7-07bc791520f2",
        "labId": "9eb49874-5cf6-4c79-a55d-2a2155bf9459",
        "nonStandardTreatment": "Yes",
        "otherGenotypeInformation": false,
        "patientAge": 60,
        "patientId": "df405f1b-30bf-4b36-bdb8-763f6de8bed2",
        "phenotypeInformationFirstLineDrugs": false,
        "phenotypeInformationOtherDrugs": false,
        "previousTbinformation": false,
        "priorTreatmentDate": "2017-05-12T20:19:53.138Z",
        "prospectiveIsolate": false,
        "recentMdrTb": "Not known",
        "siteId": "049fb71d-8157-412e-a289-e5c74cf3d1ef",
        "smear": "+",
        "smoker": "false",
        "sputumCultureConversion": "Sputum culture positive at 2-3 months",
        "sputumSmearConversion": "Not known or not done",
        "startProgrammaticContinuationTreatment": "Not known",
        "startProgrammaticTreatment": true,
        "tbProphylaxis": "Not known",
        "tbProphylaxisDate": "2016-10-17T22:25:29.319Z",
        "wgsPlatform": "Other",
        "wgsPlatformOther": "in",
        "whoOutcomeCategory": "Lost to follow-up or defaulted",
      },
      "organisation": Object {
        "id": "591ab833d1f7fd000e5083c3",
        "name": "Ward - Auer",
      },
      "owner": Object {
        "email": "Anya46@yahoo.com",
        "firstname": "Saige",
        "id": "591ab833d1f7fd000e5083f5",
        "lastname": "Marks",
        "phone": "(773) 716-1334",
      },
      "snpDistance": Object {
        "analysed": "2017-01-21T01:52:11.092Z",
        "engine": "enim",
        "experiments": Array [],
        "id": "591ab838d1f7fd000e5084f9",
        "version": "v80049",
      },
      "uploaded": "2016-10-26T17:12:17.333Z",
    },
  ],
  "total": 2,
}
`;
